packages/contracts-bedrock/test/setup/testUtils/TestUtils.sol

```diff
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { Test } from "forge-std/Test.sol";
+import {
+    IAddressCondition,
+    AddressConditionChainer,
+    AddressConditions,
+    IsPayableCondition,
+    IsNotPayableCondition,
+    IsNotPrecompileCondition,
+    IsNotForgeAddressCondition
+} from "test/setup/testUtils/conditions/AddressConditions.sol";
+import { ForbiddenAddresses, ForbiddenUints, ForbiddenInts } from "test/setup/testUtils/Forbiddens.sol";
+
+abstract contract TestUtils is Test {
+    // This function returns _addr if it satisfies all given conditions and is not forbidden,
+    // otherwise it will generate a new random address that satisfies the conditions and is not
+    // forbidden.
+    // NOTE: This function will resort to vm.assume() if it does not find a valid address within
+    // the given number of _attempts.
+    function __randomAddress(
+        address _addr,
+        AddressConditionChainer _conditions,
+        ForbiddenAddresses _forbiddenAddresses,
+        uint256 _attempts
+    )
+        internal
+        returns (address)
+    {
+        IAddressCondition[] memory conditions = _conditions.conditions();
+
+        // if _attempts is 0, vm.assume below will fail
+        bool pass = false;
+        for (uint256 i; i < _attempts; i++) {
+            if (_forbiddenAddresses.forbiddenAddresses(_addr)) {
+                _addr = vm.randomAddress();
```

Comment:
oh this is cool, I was not aware of this method, and that random inputs can come from the env rather than input args now.